### PR TITLE
youtube-music: use system electron

### PR DESCRIPTION
**Pull request by fee1-dead**
## Description of changes

bumped version to 2.1.0, made it so that it builds from source and uses the system electron instead of extracting the appimage.

~~Note that this drops macos support as we don't package electron on macos (there is electron_bin but looks like packages don't use them)~~ Nevermind, the original package didn't support macos either

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/youtube-music/default.nix [hunk 1]
```diff
@@ -1,44 +1,86 @@
-{ lib, fetchurl, appimageTools, makeWrapper }:
+{ lib
+, fetchFromGitHub
+, buildNpmPackage
+, makeWrapper
+, electron_25
+, python3
+, copyDesktopItems
```
Comment (aacebedo) [inline pkgs/applications/audio/youtube-music/default.nix:7]:
```suggestion
, copyDesktopItems
, makeDesktopItem
```

File: pkgs/applications/audio/youtube-music/default.nix [hunk 2]
```diff
@@ -1,44 +1,86 @@
...
-
-    install -m 444 \
-        -D ${appimageContents}/youtube-music.desktop \
-        -t $out/share/applications
-    substituteInPlace \
-        $out/share/applications/youtube-music.desktop \
-        --replace 'Exec=AppRun' 'Exec=${pname}'
-    cp -r ${appimageContents}/usr/share/icons $out/share
+
+  nativeBuildInputs = [ makeWrapper python3 ];
```
Comment (aacebedo) [inline pkgs/applications/audio/youtube-music/default.nix:27]:
```suggestion
  nativeBuildInputs = [ makeWrapper python3 copyDesktopItems ];
```

**Comment by aacebedo** [comment]
Had to make some changes to make it compile and also to ensure the desktop items are installed. 

I am also on wayland (sway) and I observed two problems:
- The app opens and the debug tools of chrome are automatically opened
- The app randomly crashes at startup

Do you think it comes from electron or the app itself?
I have a long list of problems with electron-based apps, really hate them.

**Comment by fee1-dead** [comment]
> The app opens and the debug tools of chrome are automatically opened

I looked a bit into this, and it looks like it is because we are running `electron` instead of the packaged binary. Setting an environment variable fixed it.

> The app randomly crashes at startup

Yeah.. vscode also randomly crashes for me. I don't know if I have a solution.

**Review by aacebedo — APPROVED** [review]

**Comment by aacebedo** [comment]
> > The app opens and the debug tools of chrome are automatically opened
> 
> I looked a bit into this, and it looks like it is because we are running `electron` instead of the packaged binary. Setting an environment variable fixed it.
> 
> > The app randomly crashes at startup
> 
> Yeah.. vscode also randomly crashes for me. I don't have a solution.

Note that for VSCode, setting  `"window.titleBarStyle": "custom"` decreases significantly the crashes on my side.

File: pkgs/applications/audio/youtube-music/default.nix [hunk 3]
```diff
@@ -1,44 +1,88 @@
...
+    })
+  ];
+
   meta = with lib; {
     description = "Electron wrapper around YouTube Music";
     homepage = "https://th-ch.github.io/youtube-music/";
     license = licenses.mit;
-    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
-    platforms = platforms.linux;
+    inherit (electron.meta) platforms;
```
[the remark this answers is not in the document]
Comment (fee1-dead) [reply pkgs/applications/audio/youtube-music/default.nix:83]: `electron` doesn't support macos.

File: pkgs/applications/audio/youtube-music/default.nix [hunk 4]
```diff
@@ -1,44 +1,88 @@
...
+  '';
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "youtube-music";
+      exec = "youtube-music %u";
+      icon = "youtube-music";
+      desktopName = "Youtube Music";
+      startupWMClass = "Youtube Music";
+      comment = meta.description;
```
[the remark this answers is not in the document]
Comment (fee1-dead) [reply pkgs/applications/audio/youtube-music/default.nix]: removed

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by khaneliman** [comment]
Result of `nixpkgs-review pr 260995` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Review by aacebedo — APPROVED** [review]

File: pkgs/applications/audio/youtube-music/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/applications/audio/youtube-music/default.nix:83]: > `electron` doesn't support macos. This package should support macos.

`electron` does support macOS, though `electron.meta.platforms` probably won't give darwin when evaluated on Linux.

[the remark this answers is not in the document]
Comment (fee1-dead) [reply pkgs/applications/audio/youtube-music/default.nix:83]: Oh huh. I thought it only supported linux given the result from search.nixos.org. I don't have a MacOS machine available to test this, if you could test this on a macos machine that would be very nice.

**Comment by wegank** [comment]
@ofborg build youtube-music youtube-music.passthru.tests

**Comment by wegank** [comment]
I guess this package should be marked as broken on darwin, as every other package calling `electron-builder` in nixpkgs.

There is a fix, though. `electronDist` should point to the folder containing a **writable** `Electron.app`, and `installPhase` should just copy the modified `YouTube Music.app` to `$out`. No `postInstall` is required.

```nix
  postBuild = ''
    cp -r ${electron}/Applications/Electron.app Electron.app
    chmod -R u+w Electron.app
    npm exec electron-builder -- \
      --dir \
      -c.electronDist=. \
      -c.electronVersion=${electron.version}
  '';
  
  installPhase = ''
    runHook preInstall

    mkdir -p $out/{Applications,bin}
    mv pack/mac-*/*.app $out/Applications
    makeWrapper $out/Applications/*.app/Contents/MacOS/* $out/bin/youtube-music

    runHook postInstall
  '';
```

**Comment by wegank** [comment]
I'll try to add darwin support in a follow-up PR. This PR currently has no darwin rebuilds according to ofborg, so should be good to merge.

**Pull request merged**